oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java

```diff
@@ -562,21 +569,88 @@ public AccessToken refreshAccessToken() throws IOException {
+        : getIdTokenIamEndpoint(targetAudience);
+  }
 
-    JsonFactory jsonFactory = OAuth2Utils.JSON_FACTORY;
+  /**
+   * Uses the Oauth Endpoint to generate an ID token. Assertions and grant_type are sent in the
+   * request body.
+   */
+  private IdToken getIdTokenOauthEndpoint(String targetAudience) throws IOException {
     long currentTime = clock.currentTimeMillis();
     String assertion =
-        createAssertionForIdToken(
-            jsonFactory, currentTime, tokenServerUri.toString(), targetAudience);
+        createAssertionForIdToken(currentTime, tokenServerUri.toString(), targetAudience);
 
+    Map<String, Object> requestParams =
+        ImmutableMap.of("grant_type", GRANT_TYPE, "assertion", assertion);
     GenericData tokenRequest = new GenericData();
-    tokenRequest.set("grant_type", GRANT_TYPE);
-    tokenRequest.set("assertion", assertion);
+    requestParams.forEach(tokenRequest::set);
+    UrlEncodedContent content = new UrlEncodedContent(tokenRequest);
+
+    HttpRequest request = buildIdTokenRequest(tokenServerUri, transportFactory, content);
+    HttpResponse httpResponse = executeRequest(request);
+
+    GenericData responseData = httpResponse.parseAs(GenericData.class);
+    String rawToken = OAuth2Utils.validateString(responseData, "id_token", PARSE_ERROR_PREFIX);
+    return IdToken.create(rawToken);
+  }
+
+  /**
+   * Use IAM generateIdToken endpoint to obtain an ID token.
+   *
+   * <p>This flow works as follows:
+   *
+   * <ol>
+   *   <li>Create a self-signed jwt with `https://www.googleapis.com/auth/iam` as the scope.
+   *   <li>Use the self-signed jwt as the access token, and make a POST request to IAM
+   *       generateIdToken endpoint.
+   *   <li>If the request is successfully, it will return {"token":"the ID token"}. Extract the ID
+   *       token.
+   * </ol>
+   */
+  private IdToken getIdTokenIamEndpoint(String targetAudience) throws IOException {
+    JwtCredentials selfSignedJwtCredentials =
+        createSelfSignedJwtCredentials(
+            null, ImmutableList.of("https://www.googleapis.com/auth/iam"));
+    Map<String, List<String>> responseMetadata = selfSignedJwtCredentials.getRequestMetadata(null);
+    // JwtCredentials will return a map with one entry ("Authorization" -> List with size 1)
+    String accessToken = responseMetadata.get(AuthHttpConstants.AUTHORIZATION).get(0);
+
+    // Do not check user options. These params are always set regardless of options configured
+    Map<String, Object> requestParams =
+        ImmutableMap.of("audience", targetAudience, "includeEmail", "true", "useEmailAzp", "true");
+    GenericData tokenRequest = new GenericData();
+    requestParams.forEach(tokenRequest::set);
     UrlEncodedContent content = new UrlEncodedContent(tokenRequest);
 
+    URI iamIdTokenUri =
+        URI.create(
+            String.format(
+                OAuth2Utils.IAM_ID_TOKEN_ENDPOINT_FORMAT, getUniverseDomain(), clientEmail));
```

Comment:
nit: Can we be consistent with `iamIdTokenUri` and `tokenServerUri`? either both declare as class variable or both inside respective get endpoint method.

Reply:
tokenServerUri is a customizable from setters and can't be declared as a class variable. `iamIdTokenUri` probably can be initialized in the constructor and I'll move it up there.

Reply:
On second thought, `getUniverseDomain()` throws an IOException and adding a checked exception to the constructor is a breaking change. I think it would be better to leave it here.